infra/Pos/Slotting/Types.hs

```diff
@@ -23,14 +33,56 @@ data EpochSlottingData = EpochSlottingData
 instance NFData EpochSlottingData
 
 -- | Data necessary for slotting to work which is basically part of GState.
-data SlottingData = SlottingData
-    { sdPenult      :: !EpochSlottingData
-    -- ^ 'EpochSlottingData' for penult (i. e. 'sdPenultEpoch') epoch.
-    , sdLast        :: !EpochSlottingData
-    -- ^ 'EpochSlottingData' for last epoch.
-    -- Last means 'sdPenultEpoch + 1'.
-    , sdPenultEpoch :: !EpochIndex
-    -- ^ Penult epoch for which 'EpochSlottingData' is known.
-    } deriving (Eq, Show, Generic)
+-- External code can use functions like getLastEpochIndex or getLastEpochSlottingData and 
+-- not worry about cases where it doesn't exist (this module should be responsible for it).
+-- Note that it's important to use error rather than default values like 0, because 
+-- such cases indicate invariants violation and shouldn't be hidden behind default values.
+type SlottingData = Map EpochIndex EpochSlottingData
+
+-- | Get the latest epoch index.
+getLastEpochIndex :: SlottingData -> EpochIndex
+getLastEpochIndex = fst . M.findMax
+
+-- | Get the latest epoch slotting data.
+getLastEpoch :: SlottingData -> EpochSlottingData
```

Comment:
I would call it `getLastEpochData` or `getLastEpochSlottingData`. Otherwise it's not obvious from the name that it returns `EpochSlottingData` and not `EpochIndex` (if you don't see this module).